service/kvmclusterconfig/v1/resource/encryptionkey/update.go

```diff
@@ -16,5 +17,30 @@ func (r *Resource) ApplyUpdateChange(ctx context.Context, obj, updateChange inte
 // NewUpdatePatch computes appropriate Patch based on difference in current
 // state and desired state.
 func (r *Resource) NewUpdatePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*framework.Patch, error) {
+	create, err := r.newCreateChange(ctx, obj, currentState, desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	update, err := r.newUpdateChange(ctx, obj, currentState, desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	patch := framework.NewPatch()
+
+	if create != nil {
+		patch.SetCreateChange(create)
+	}
+
+	if update != nil {
+		patch.SetUpdateChange(update)
+	}
```

Comment:
Well no, there must not be any conditionals in here. The patch actually sorts and decides what to do with the receiving change. We should not care about its actual state here. It is totally fine to pass a nil change because this leads to the "apply" methods being called and they can decide what to do with the information. Maybe they want to emit some logs or metrics if the change is nil, which is mostly true for all operator implementations as of now.

Reply:
Patch doesn't have any means for verifying whether it has some specific change or not, making it currently impossible to test. With `reflect.DeepEqual` it's currently "possible" without exposing Patch internals in API, but it doesn't work without these conditions. I would like to test the resource implementation. I understand that in this specific case the implementation is quite simple and there's not much to test, but it's not necessarily always like that.

Do you have suggestions how to write unit tests for the functions without writing reflection logic for Patch internals verification?

Reply:
I do not understand the problem here. What do you want to test? You can test `ApplyCreateChange` for instance and `newCreateChange`. So far there was actually no real reason to test `NewUpdatePatch`. There is no real logic in there and once written down it never changes so I miss the point for unit tests here. I am not a fan of 100% coverage because this is just not forth it most of the time. I consider this here testing the last 5% but we should ask if it is just worth it, and I would just say no, no it is not. It does not make that much sense to test if the patch implementation works and making sure two methods are called which is not likely to change in the future is not worth it I think. At the end of the day there should also be tests and QA on other levels than unit tests to cover certain things unit tests cannot accomplish.

Reply:
You are right. I got too fixated to initial tests I wrote for `NewUpdatePatch` and didn't realize the aspect of testing sub-routines separately. With a bit of typing that works. I'll refactor the tests now.

One thing I still disagree:I think that `framework.Patch` should provide testing aid because it affects testability of `NewUpdatePatch` implementation and putting extra indirection there is a matter of taste. Not every library user wants to do that.

Reply:
When you think this is important get in touch with Pawel because he designed the patch magic and probably had reasons for certain things, or maybe not. Uncertain. Anyway, if there is a need we can just adapt and improve the situation. So yeah, thanks for pushing this buddy.

Reply:
When you have `newCreateChange` extracted the only thing you can test is `patch.SetCreateChange` call. Eventually you can test one `if` with another `if` which (arguably) is like writing the same code twice.

You can check if `ApplyCreateChange` was called after `patch.SetCreateChange` but that would be verification if operatorkit implementation is correct. Which is/should be done in the operatorkit tests. Usually users trust libraries they use.

Question is if we want to sacrifice clean API (introduce confusion) for that little coverage?